app/services/search_indexer.rb

```diff
@@ -149,4 +225,18 @@ def characters(string)
       scrubbed << " "
     end
   end
+
+  private
+
+  def self.migrate_core_version
+    if SiteSetting.search_index_core_version < 1
+      if ['zh_TW', 'zh_CN', 'ja', 'ko'].include?(SiteSetting.default_locale) || SiteSetting.search_tokenize_chinese_japanese_korean
+        self.force_reindex_posts_topics
+        self.force_reindex_categories
+        self.force_reindex_users
+      end
+
+      SiteSetting.search_index_core_version = 1
+    end
+  end
```

Comment:
this should be in a once off job, I don't particularly like having this in the core class.

version should be stored in the search table and then we can have a regular weekly job that forces are reindex on posts that have an earlier version and cleanly recover from partial re-index

note I am going to be out for a week, so it is going to take a while to merge this in.